Return f32 and f64 in XMM0 instead of FP0 on i686

GuentherVIII opened the pull request:
i686 already uses SSE to do calculations with f32 and f64, but the C calling convention uses the x87 stack to return values. The Rust calling convention does not need to do this, and LLVM makes it easy to use XMM0 instead, which saves move instructions and fixes problems with NaN values.

See https://github.com/rust-lang/rust/issues/115567, https://github.com/rust-lang/rust/issues/91399#issuecomment-1045989643 and [this LLVM comment](https://github.com/rust-lang/llvm-project/blob/0537f6354cffe546cbf47f6dc9c7f82e49e86cfb/llvm/lib/Target/X86/X86CallingConv.td#L269).

Commit:
Return f32 and f64 in XMM0 instead of FP0 on i686 Rust calling convention

i686 already uses SSE2 to do calculations with f32 and f64, but the C
calling convention uses the x87 stack to return values. The Rust calling
convention does not need to do this, and LLVM makes it easy to use XMM0
instead, which saves move instructions and fixes problems with NaN values.

the8472 commented on library/std/src/f32/tests.rs:
```diff
@@ -326,9 +325,6 @@ macro_rules! assert_f32_biteq {
     };
 }
 
-// Ignore test on x87 floating point, these platforms do not guarantee NaN
-// payloads are preserved and flush denormals to zero, failing the tests.
-#[cfg(not(target_arch = "x86"))]
```
I think the tests still should be ignored on pre-SSE x86 targets. We have a bunch of i586 tier2 targets and others might be running tests for them.

GuentherVIII replied:
Done. Running the i586 tests also required changes to rustc_ty_utils which I'm not happy with. I could not figure out how to properly detect SSE2 support - target.features is empty on i586-unknown-linux-gnu at that point.

GuentherVIII commented:
@rustbot label +A-abi +A-floating-point +O-x86 -T-libs

chorman0773 commented:
Can I ask that we not make floating-point ABI *even* more complicated on x86?

chorman0773 commented:
@rust-lang/opsem and probably @rust-lang/lang definitely needs to look at this for the reason in #116344.

When compiling for i586, local `#[target_feature]` attributes can likely affect whether an `f32` or `f64` return value is passed in an SSE or x87 register according to the code from this PR from what I can tell. The setting for the callee is *not* respected for the caller, at least in the existing cases affecting by this. This further constrains what can be promised in #115476.

workingjubilee commented:
> Can I ask that we not make floating-point ABI even more complicated on x86?
Rustc already has ABI issues with floats and -x87, adding +sse to the list of features replied upon for abi is just going to make that problem worse.
>
> Not to mention my headache trying to catalog all of the different x86 float ABIs.

Given that we are also discussing making SSE/SSE2 mandatory for i686 targets with SSE2 as part of their baseline features, I think we should forge ahead, because this actually makes it more like x86-64's ABI.

programmerjake commented on compiler/rustc_ty_utils/src/abi.rs:
```diff
@@ -371,6 +371,8 @@ fn fn_abi_new_uncached<'tcx>(
     let target_env_gnu_like = matches!(&target.env[..], "gnu" | "musl" | "uclibc");
     let win_x64_gnu = target.os == "windows" && target.arch == "x86_64" && target.env == "gnu";
+    let x86_sse2 =
+        target.arch == "x86" && (target.features.contains("sse2") || target.cpu == "pentium4");
```
I expect you'd want to query the `sse2` target feature rather than checking for the `pentium4` cpu, since e.g. `penryn` also supports `sse2` but isn't the `pentium4`. https://rust.godbolt.org/z/j7oPE6Wrf

also, what if there's a soft-float target for an OS kernel where `sse2` would be disabled even if the cpu is `pentium4` due to the kernel not saving SSE registers?

workingjubilee replied:
@GuentherVIII If you handle the aforementioned edge-cases appropriately, I believe we can get this approved. We want to only do this if the target's baseline features include SSE2.

programmerjake replied:
so, in other words, only if `SSE2` is implied by the target structs/json ignoring any `-C target-features` or `-C target-cpu` or other overrides.

GuentherVIII replied:
Does rust currently ask LLVM for the target features implied by the target cpu? I could only find the code that creates the list for `#[cfg(target_feature)]`, which takes command-line options into account.

I've changed the implementation to check that list instead of target.features and target.cpu. This has the advantage that the standard library can activate the NAN-returning test precisely when it passes, and deactivating SSE with -C target-features still works.

There's also a second commit introducing yet another target option to make this configurable in the target specification and enabling it in all "i686" targets except i686_unknown_uefi, which disables SSE. The standard

I do not think we should make the ABI dependant on `#[target_feature]` - unlike -C target-feature, the attribute cannot disable SSE. If someone disables SSE with -C target-feature and re-enables it for select functions with `#[target_feature]`, this shouldn't result in incompatible function ABIs with the same function type.

programmerjake replied:
the issue you'll run into with `-C target-feature=-sse` is that functions in your binary won't match the ABI of other code (such as `std`) with SSE2 enabled in the default target features

the8472 replied:
There's talk about forbidding going below the baseline. If you want less-than-baseline features you need a different target.

https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/Float.20ABI.20hell/near/394473707

RalfJung commented:
> Rustc already has ABI issues with floats and -x87, adding +sse to the list of features replied upon for abi is just going to make that problem worse.

I think it just means we want to have both of these in the set of baseline features.

But it means we should maybe block this PR on having the concept of "baseline features" rolled out so people get at least a warning when they disable a feature that should not be disabled?

thomcc commented:
r? compiler

Commit:
Enable usage of XMM0 per-target

Activating SSE2 with -C target-feature doesn't change the Rust ABI anymore,
while deactivating it still does.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6444605590/job/17497781942) [(plain)](https://github.com/rust-lang/rust/commit/2b153e7262703fad18d4d5791062db48f82c27a5/checks/17497781942/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=GuentherVIII
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_5c14986a-fa5a-445c-aa45-d9a6ea39efb2
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=96df6bb68a28d1a4f9c62417ff793a58fa3cfeff
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_5c14986a-fa5a-445c-aa45-d9a6ea39efb2
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_5c14986a-fa5a-445c-aa45-d9a6ea39efb2
GITHUB_TRIGGERING_ACTOR=GuentherVIII
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115919/merge
GITHUB_WORKFLOW_SHA=96df6bb68a28d1a4f9c62417ff793a58fa3cfeff
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---

     Running unittests src/lib.rs (obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_target-d139d242fe22457e)

running 228 tests
....................................................................................FFFF  88/228
.FFFFF.FFFFFFF...F...................................................................... 176/228

failures:

---- spec::tests::i586_unknown_linux_gnu stdout ----
---- spec::tests::i586_unknown_linux_gnu stdout ----
thread 'spec::tests::i586_unknown_linux_gnu' panicked at compiler/rustc_target/src/spec/tests/tests_impl.rs:9:5:
```

</details>

b-naber commented:
r? compiler

petrochenkov commented:
This should probably be blocked on https://github.com/rust-lang/rust/pull/116584.
@rustbot blocked

bors commented:
:umbrella: The latest upstream changes (presumably #117716) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
Closing because https://github.com/rust-lang/rust/pull/116584 was closed.

RalfJung commented:
This got superseded by https://github.com/rust-lang/rust/pull/123351 which I think had the same effect.

programmerjake commented:
> This got superseded by #123351 which avoids the ABI issue by returning things in-memory.

returning stuff in xmm0 is still better since it avoids extra load/store, so I think we still want this PR (or similar) but it's not as critical since the corrupting-float-returns bug is now fixed.
